### PR TITLE
feat(auto-balance): disk in pressure

### PR DESCRIPTION
**Pull request by c3y1huang**
#### Which issue(s) this PR fixes:

Issue longhorn/longhorn#4105

#### What this PR does / why we need it:

- Introduces new `replica-auto-balance-disk-pressure-percentage` setting.
- Implement local file sync for newly added replicas.

#### Special notes for your reviewer:

`None`

#### Additional documentation or context

`None`

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -34,6 +34,8 @@ replace (
 	k8s.io/sample-controller => k8s.io/sample-controller v0.28.5
 )
 
+replace github.com/longhorn/longhorn-engine => github.com/c3y1huang/longhorn-engine v1.6.0-dev-20240105.0.20240201041530-b0e5b9da243c
```
Comment (c3y1huang) [inline go.mod:37]: TODO: update after https://github.com/longhorn/longhorn-engine/pull/1003.

File: go.mod [hunk 2]
```diff
@@ -36,6 +36,8 @@ replace (
 
 replace github.com/longhorn/longhorn-engine => github.com/c3y1huang/longhorn-engine v1.6.0-dev-20240105.0.20240201041530-b0e5b9da243c
 
+replace github.com/longhorn/longhorn-instance-manager => github.com/c3y1huang/longhorn-instance-manager v1.6.0-dev-20240105.0.20240201065454-cfeb74abb194
```
Comment (c3y1huang) [inline go.mod:39]: TODO: update after https://github.com/longhorn/longhorn-instance-manager/pull/390.

File: controller/engine_controller.go [hunk 3]
```diff
@@ -1909,17 +1910,22 @@ func (ec *EngineController) startRebuilding(e *longhorn.Engine, replicaName, add
 			return
 		}
 
+		localSync, err := ec.getFileLocalSync(replica, e)
+		if err != nil {
+			ec.logger.WithError(err).Errorf("Failed to initiate file local sync for replica %v, use remote sync", replicaName)
```
Comment (derekbit) [inline controller/engine_controller.go:1915]:
```suggestion
			ec.logger.WithError(err).Warnf("Failed to initiate file local sync for replica %v, use remote sync", replicaName)
```

File: controller/volume_controller.go [hunk 4]
```diff
@@ -2131,6 +2106,64 @@ func (c *VolumeController) getPreferredReplicaCandidatesForDeletion(rs map[strin
 	return deletionCandidates, nil
 }
 
+// getSortedReplicasByStorageAvailable retrieves a sorted list of replica names
+// based on the storage available of the disk the replica is located.
+// If any error occurs, the function wraps the error with additional information
+// and returns it.
+func (c *VolumeController) getSortedReplicasByStorageAvailable(replicaCandidates []string, replicas map[string]*longhorn.Replica) ([]string, error) {
```
Comment (derekbit) [inline controller/volume_controller.go:2113]: nit: rename to `getSortedReplicasByAscendingStorageAvailable`

File: controller/volume_controller.go [hunk 5]
```diff
@@ -2358,14 +2394,197 @@ func (c *VolumeController) getReplicaCountForAutoBalanceBestEffort(v *longhorn.V
...
-		return 0, nil, []string{}
+		mostExtraRList = nil
+		leastExtraROwners = []string{}
+	} else {
+		adjustCount = mostExtraRCount - leastExtraRCount - 1
+		log.Infof("Found %v replicas from %v to balance to one of node in %v", adjustCount, mostExtraRList, leastExtraROwners)
+	}
+
+	if adjustCount == 0 {
+		replicasUnderDiskPressure, err := c.getReplicasUnderDiskPressure()
```
Comment (derekbit) [inline controller/volume_controller.go:2405]: if `replica-soft-anti-affinity` is `disabled`, is `replicasUnderDiskPressure` be an empty map?

File: controller/volume_controller.go (hunk 4 above)
Comment (c3y1huang) [reply controller/volume_controller.go:2113]: I am not sure about this. The `getSortedReplicasByAscendingStorageAvailable` looks like we are storing the replicas by available size, but we are actually sorting them by name.

File: controller/volume_controller.go (hunk 5 above)
Comment (c3y1huang) [reply controller/volume_controller.go:2412]: I am also fine with removing the dependency on `replica-soft-anti-affinity` setting.

https://github.com/longhorn/longhorn/pull/7576#issuecomment-2190479295

File: controller/volume_controller.go (hunk 4 above)
Comment (derekbit) [reply controller/volume_controller.go:2113]: `replicaMapByStorage[storageAvailable] = append(replicaMapByStorage[storageAvailable], replica.Name)`
The key is storageAvailable, and the value is the replica name. Shouldn't the list be sorted in ascending order of storageAvailable after using util.SortKeys()?

Comment (c3y1huang) [reply controller/volume_controller.go:2113]: Oh yes, you're right.

**Comment by c3y1huang** [comment]
@derekbit this is ready for review.

File: controller/engine_controller.go [hunk 6]
```diff
@@ -1998,6 +2004,52 @@ func (ec *EngineController) startRebuilding(e *longhorn.Engine, replicaName, add
...
+		return nil, err
+	}
+
+	// Check if there are multiple replicas on the target replica's node,
+	// do nothing if there is no other available replica to sync locally from.
+	if len(replicaMapByNode[targetReplica.Spec.InstanceSpec.NodeID]) <= 1 {
+		return nil, nil
+	}
+
+	localSync := &etypes.FileLocalSync{}
```
Comment (derekbit) [inline controller/engine_controller.go:2024]:
```suggestion
	var localSync *etypes.FileLocalSync
```
`localSync` is initialized to a non-nil value.
If the following for-loop cannot find a valid replica, the non-nil is returned and without error.

**Review by derekbit — APPROVED** [review]
LGTM

**Comment by derekbit** [comment]
If @shuo-wu and @innobead has no other concerns, we can merge it.

**Comment by derekbit** [comment]
We can merge it after the conflicts are resolved.

File: datastore/longhorn.go [hunk 7]
```diff
@@ -656,6 +656,32 @@ func (s *DataStore) ListSettings() (map[types.SettingName]*longhorn.Setting, err
 	return itemMap, nil
 }
 
+func (s *DataStore) GetAutoBalancedReplicasSetting(volume *longhorn.Volume) (longhorn.ReplicaAutoBalance, error) {
```
Comment (derekbit) [inline datastore/longhorn.go:659]: nit: add a doc for the function

File: scheduler/replica_scheduler.go [hunk 8]
```diff
@@ -792,6 +795,32 @@ func (rcs *ReplicaScheduler) IsSchedulableToDisk(size int64, requiredStorage int
...
+		"diskUUID":               info.DiskUUID,
+		"diskPressurePercentage": diskPressurePercentage,
+		"requiredStorage":        requiredStorage,
+		"storageScheduled":       info.StorageScheduled,
+		"storageReserved":        info.StorageReserved,
+		"storageMaximum":         info.StorageMaximum,
+	}).Debugf("Evaluated new disk usage percentage after scheduling replica: %v%%", newDiskUsagePercentage)
+
+	return rcs.IsSchedulableToDisk(size, requiredStorage, info) &&
+		newDiskUsagePercentage < int64(diskPressurePercentage)
```
Comment (derekbit) [inline scheduler/replica_scheduler.go:810]: Like the split that makes the logic clear.
We can have a code convention rules and checks in the future.
https://go.dev/wiki/CodeReviewComments

cc @innobead

File: datastore/longhorn.go [hunk 9]
```diff
@@ -656,6 +656,44 @@ func (s *DataStore) ListSettings() (map[types.SettingName]*longhorn.Setting, err
...
+		}
+
+		setting = longhorn.ReplicaAutoBalance(globalSetting)
+	}
+
+	err = types.ValidateReplicaAutoBalance(longhorn.ReplicaAutoBalance(setting))
+	if err != nil {
+		setting = longhorn.ReplicaAutoBalanceDisabled
+	}
+	return setting, errors.Wrapf(err, "replica auto-balance is disabled")
```
Comment (derekbit) [inline datastore/longhorn.go:694]: Should we return error `errors.Wrapf(err, "replica auto-balance is disabled")`?

Comment (c3y1huang) [reply datastore/longhorn.go]: The original purpose was to let the caller decide how to handle the error. It seems unnecessary to return this; we can refactor to consolidate the logging within this function, making it cleaner. Thank you, and updated.

**Review by derekbit — APPROVED** [review]

**Pull request merged**